File: crates/fey_math/src/projection.rs
use crate::{Num, impl_approx, impl_bytemuck, impl_casts, impl_serde, impl_tuple_arr};

pub type ProjectionF = Projection<f32>;

/// Represents the projection of a 2D shape on an axis.
///
/// This primitive does not contain the axis itself, merely
/// the start and end bounds of the projection. This is used
/// in overlap checks for convex shapes.
#[repr(C)]
#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
pub struct Projection<T> {
    pub min: T,
    pub max: T,
}

impl_bytemuck!(Projection);

impl_tuple_arr!(
    NAME = Projection
    LEN = 2
    FIELDS = (min, max)
    TUPLE = (T, T)
);

impl_approx!(
    NAME = Projection
    FIELDS = (min, max)
);

impl_serde!(
    NAME = Projection
    FIELDS = (min, max)
);

impl_casts!(
    NAME = Projection
    FIELDS = (min, max)
);

impl<T> Projection<T> {
    /// Create a new projection.
    #[inline]
    pub const fn new(min: T, max: T) -> Self {
        Self { min, max }
    }
}

impl<T: Num> Projection<T> {
    /// Returns true if this projection overlaps the other.
    #[inline]
    pub fn overlaps(&self, other: Projection<T>) -> bool {
        self.min < other.max && self.max > other.min
    }

    /// If this projection overlaps the other, returns the amount
    /// which it overlaps.
    #[inline]
    pub fn overlap(&self, other: Projection<T>) -> Option<T> {
        (self.min < other.max && self.max > other.min).then(|| self.max - other.min)
    }

    /// Length of the projection.
    #[inline]
    pub fn len(&self) -> T {
        self.max - self.min
    }
}

File: crates/fey_math/lua/Projection.lua
---@meta

---Represents the projection of a 2D shape onto an axis.
---@class (exact) Projection: ProjectionMethods
---@field min number
---@field max number

---@class ProjectionClass : ProjectionMethods
---@overload fun(min: number, max: number): Projection
local module = {}

---@class ProjectionMethods
local methods = {}

---Create a new Projection.
---@param min number
---@param max number
---@return Projection
---@nodiscard
function module.new(min, max) end

---Returns a temporary copy of this value.
---@param self Projection
---@return Projection
---@nodiscard
function methods.clone(self) end

---Boxes the value if not already boxed.
---@param self Projection
---@return Projection
---@nodiscard
function methods.box(self) end

---Boxes a copy of this value.
---@param self Projection
---@return Projection
---@nodiscard
function methods.box_clone(self) end

---Returns `true` if the two projections are approximately equal.
---@param self Projection
---@param other Projection
---@return boolean
---@nodiscard
function methods.approx(self, other) end

---Length of the projection `|max - min|`.
---@param self Projection
---@return number
---@nodiscard
function methods.len(self) end

---If this projection overlaps the other, returns the amount which it overlaps.
---@param self Projection
---@param other Projection
---@return number?
---@nodiscard
function methods.overlap(self, other) end

---Returns `true` if the projections overlap.
---@param self Projection
---@param other Projection
---@return boolean
---@nodiscard
function methods.overlaps(self, other) end

return module